clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4313,6 +4325,18 @@ bool Util::isSyclKernelHandlerType(const QualType &Ty) {
   return matchQualifiedTypeName(Ty, Scopes);
 }
 
+bool Util::isSyclAccessorNoAliasPropertyType(QualType Ty) {
+  std::array<DeclContextDesc, 6> Scopes = {
+      Util::DeclContextDesc{Decl::Kind::Namespace, "cl"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "sycl"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "ONEAPI"},
+      Util::DeclContextDesc{Decl::Kind::Namespace, "property"},
+      Util::DeclContextDesc{Decl::Kind::CXXRecord, "no_alias"},
+      Util::DeclContextDesc{Decl::Kind::ClassTemplateSpecialization,
+                            "instance"}};
```

Comment:
I have to ask this to sleep better at night. 
Do we use this method because there is no way to write a partial specialization for the accessor class, which uses `nolias` attribute?

I think we are trying to avoid such runtime library declarations matchers and use more robust approach with a new attribute - https://github.com/intel/llvm/pull/2091. I would expect this `property` will be moved from `ONEAPI` to `sycl` namespace in the future and it would be great to avoid compiler changes for such library API updates.

Reply:
@bader The idea behind #2091 is not fully clear to me i.e., I am not sure how this idea could be applied to accessor properties. Could you please elaborate? 

At the moment, the way to understand each of the accessor properties(buffer_location @MrSidims, noalias etc.,) seems, in a straightforward way, to just check for its name and handle it based on that, since there is no indication in the spec regarding the change of scope. Are you suggesting adding an attribute to identify/indicate the accessor property parameter?

Reply:
@bader 
I understand the merit of avoiding the fragility caused by relying on SYCL header scope gathered from the DeclContext; This needs to be addressed for not only no_alias, but also buffer_location and other (future) accessor properties such as no_offset and noinit. This can be addressed using an attribute consistently for all the accessor properties. I can create a follow-up task/tracker to resolve this issue separately. Please let me know your thoughts. Thank you.

Reply:
I agree with @bader that this is a fragile approach that ties the FE very tightly to the SYCL header files, but I don't think there's anything that can be done about it as part of this patch and so no changes are needed. Once we have the better way available to us, we should hit up all of these properties at once. @bader, please correct me if I'm wrong.

Reply:
I agree, I just wanted to check if anyone has explored an option to support `no_alias` accessor property by specializing accessor class template to re-use existing `noalias` attribute.

Reply:
https://github.com/intel/llvm/issues/3529 - follow up issue

Reply:
> Would you suggest adding an attribute to identify/indicate the accessor property parameter, maybe at a later time?

According to my understanding this patch applies `noalias` attribute to an accessor pointer member when accessor is instantiated with special template parameter. `noalias` attribute is already exposed to source level, so I'm wondering why can't we just apply it in DPC++ headers instead? This way, no compiler changes are required (maybe except propagating attribute to the generated kernel function parameters).
I guess someone from the runtime team should be able to answer this. @romanovvlad, do you know who can help to clarify this?